File: string__reverse_words_inplace.js
function reverseString(str, s, e) {
    if (s === undefined) s = 0;
    if (e === undefined) e = str.length - 1;
    while (s < e) {
        [str[s], str[e]] = [str[e], str[s]];
        s++;
        e--;
    }
}

function reverseWords(str) {
    var start = 0;
    var end = 0;
    var length = str.length;
    while (end < length) {
        if (str[end] !== ' ') {
            start = end;
            while (str[end] !== ' ' && end < length) end++;
            end--;
            reverseString(str, start, end);
        }
        end++;
    }
}

str = 'foobar bazbat';
console.log(str);
str = str.split('');
reverseString(str);
reverseWords(str);
console.log(str.join(''));

File: string__reverse_words_inplace.py
def reverseString(str, start=0, end=None):
    if end == None: 
        end = len(str) - 1
    while start < end:
        str[start], str[end] = str[end], str[start]
        start += 1
        end -= 1

def reverseWords(str):
    start = 0
    end = 0
    length = len(str)
    while end < length:
        if str[end] != ' ':
            start = end
            while end < length and str[end] != ' ': 
                end += 1
            end -= 1
            reverseString(str, start, end)
        end += 1

str = 'Do or do not, there is no try.'
str = list(str)
reverseString(str)
reverseWords(str)
print(''.join(str))
